lumicks/pylake/confocal.py

```diff
@@ -0,0 +1,231 @@
+import json
+import numpy as np
+
+from .detail.mixin import PhotonCounts
+from .detail.mixin import ExcitationLaserPower
+from .detail.image import save_tiff, ImageMetadata
+
+
+class BaseScan(PhotonCounts, ExcitationLaserPower):
+    """Base class for confocal scans
+
+    Parameters
+    ----------
+    name : str
+        confocal scan name
+    file : lumicks.pylake.File
+        Parent file. Contains the channel data.
+    start : int
+        Start point in the relevant info wave.
+    stop : int
+        End point in the relevant info wave.
+    json : dict
+        Dictionary containing metadata.
+    """
+
+    def __init__(self, name, file, start, stop, json):
+        self.start = start
+        self.stop = stop
+        self.name = name
+        self.json = json
+        self.file = file
+        self._cache = {}
+
+    @classmethod
+    def from_dataset(cls, h5py_dset, file):
+        """
+        Construct a confocal class from dataset.
+
+        Parameters
+        ----------
+        h5py_dset : h5py.Dataset
+            The original HDF5 dataset containing confocal scan information
+        file : lumicks.pylake.File
+            The parent file. Used to loop up channel data
+        """
+        start = h5py_dset.attrs["Start time (ns)"]
+        stop = h5py_dset.attrs["Stop time (ns)"]
+        name = h5py_dset.name.split("/")[-1]
+        try:
+            json_data = json.loads(h5py_dset[()])["value0"]
+        except KeyError:
+            # TODO => use class name
+            raise KeyError(f"Scan '{name}' is missing metadata and cannot be loaded")
+
+        return cls(name, file, start, stop, json_data)
+
+    def _get_photon_count(self, name):
+        timeline_start = self.file.h5["Photon count"][name].attrs["Start time (ns)"]
+        timeline_sample_rate = self.file.h5["Photon count"][name].attrs["Sample rate (Hz)"]
+        timeline_dt = int(1e9 /timeline_sample_rate)
+
+        # Workaround for a bug in the STED delay mechanism which could result in scan start times ending up within
+        # the sample time.
+        if timeline_start - timeline_dt < self.start < timeline_start:
+            self.start = timeline_start
```

Comment:
This is not correct. The timeline can start well before the `Scan` and in most use cases does. Pre-refactor the timeline was sliced, after which you'd get the appropriate start from `Slice.start` (it being the first timestamp after `Scan.start`).

However, for the STED thing, there was an error that could sometimes lead to a small temporal displacement of the `Scan` timestamp that wasn't quantized to whole timestamp. When you slice the timeline then, you will end up with the next sample that's valid, but it won't line up with the `Scan.start`. The reason the test didn't pick this up is because the test didn't incorporate data prior to the `Scan`. My bad on the test really.

Please have a peek at branch `stedbug_redux`. I've put a hardened test there which should pick up this regression. Would be nice to include it.